File: training-data/ocaml/max_subarray.ml
let kadane lst =
  match lst with
  | [] -> 0
  | h :: t ->
      let best, _ =
        List.fold_left
          (fun (best, current) x ->
            let new_current = max x (current + x) in
            (max best new_current, new_current))
          (h, h) t
      in
      best

let () = print_int (kadane [ -2; 1; -3; 4; -1; 2; 1; -5; 4 ]); print_newline ()

File: training-data/perl/max_subarray.pl
use strict;
use warnings;

sub kadane {
    my @items = @_;
    my $best = $items[0];
    my $current = $items[0];
    for my $i (1 .. $#items) {
        $current = $items[$i] > $current + $items[$i] ? $items[$i] : $current + $items[$i];
        $best = $current if $current > $best;
    }
    return $best;
}

print kadane(-2, 1, -3, 4, -1, 2, 1, -5, 4), "\n";
